### PR TITLE
Fixed #548: Prepare a more generic auditing table

### LINKED ISSUE #548: Prepare a more generic auditing table
Currently, we are auditing signatures. However, this may not be the best granularity since we may need to have a better audit trail across all events that occur in the system. As such, we should prepare generic auditing support, probably by integrating some well-known technologies, such as Elastic.

### PR DESCRIPTION
**Pull request by petrdvorak**

### CODE REVIEW

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/v3/OperationServiceBehavior.java [hunk 1]
```diff
@@ -399,6 +499,18 @@ public OperationDetailResponse cancelOperation(OperationCancelRequest request) t
 
         final OperationEntity savedEntity = operationRepository.save(operationEntity);
         behavior.getCallbackUrlBehavior().notifyCallbackListenersOnOperationChange(savedEntity);
+
+        logger.info("Operation canceled via  explicit server call for operation ID: {}.", operationId);
```
Comment (romanstrobl) [inline powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/v3/OperationServiceBehavior.java:503]: Extra space in message in `via  explicit`.

**Review by romanstrobl — COMMENTED** [review]
I think this is a nice addition. I would consider using the auditing library for signature verification (both online and offline) and for activations, so that all audit records are in one place? I know we have existing DB tables for signatures and activations, however some details are getting lost in them and we still need to request logs. This could be a separate ticket.

File: docs/powerauth-server/PowerAuth-Server-1.3.0.md [hunk 2]
```diff
@@ -106,6 +106,147 @@ CREATE UNIQUE INDEX pa_application_name_index
...
+- `audit_log` - table used to store audit logs
+- `audit_param` - table used to store detailed parameters for audit logs
+
+### Oracle
+
+```sql
+--
+-- Create audit log table.
+--
+CREATE TABLE audit_log (
```
Comment (banterCZ) [inline docs/powerauth-server/PowerAuth-Server-1.3.0.md:121]: `arch_db_structure.png` will be updated in a separate PR?

File: docs/powerauth-server/PowerAuth-Server-1.3.0.md [hunk 3]
```diff
@@ -106,6 +106,147 @@ CREATE UNIQUE INDEX pa_application_name_index
...
+    audit_log_id       VARCHAR2(36 CHAR) PRIMARY KEY,
+    application_name   VARCHAR2(256 CHAR) NOT NULL,
+    audit_level        VARCHAR2(32 CHAR) NOT NULL,
+    audit_type         VARCHAR2(256 CHAR),
+    timestamp_created  TIMESTAMP,
+    message            CLOB NOT NULL,
+    exception_message  CLOB,
+    stack_trace        CLOB,
+    param              CLOB,
+    calling_class      VARCHAR2(256 CHAR) NOT NULL,
```
Comment (banterCZ) [inline docs/powerauth-server/PowerAuth-Server-1.3.0.md:131]: Just wondering whether 256 chars is enough, package name + class + proxy marker... But I guess it is ok.

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/configuration/PowerAuthAuditConfiguration.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+import org.springframework.context.annotation.ComponentScan;
+import org.springframework.context.annotation.Configuration;
+
+/**
+ * Configuration of auditing for PowerAuth.
+ *
+ * @author Roman Strobl, roman.strobl@wultra.com
+ */
+@Configuration
+@ComponentScan(basePackages = {"com.wultra.core.audit.base"})
```
Comment (banterCZ) [inline powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/configuration/PowerAuthAuditConfiguration.java:33]: ComponentScan might be quite heavy. Let's discuss it later. I have opened an issue for me #726

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/v3/AuditingServiceBehavior.java [hunk 5]
```diff
@@ -51,10 +54,35 @@ public class AuditingServiceBehavior {
...
+    }
+
+    /**
+     * Log information with specified level, message, audit details, and message args.
+     * @param level Level
+     * @param message Message
+     * @param auditDetail Audit detail
+     * @param args Arguments
+     */
+    public void log(AuditLevel level, String message, AuditDetail auditDetail,  Object... args) {
```
Comment (banterCZ) [inline powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/v3/AuditingServiceBehavior.java:74]: What about to have a shortcut `logInfo(String message, AuditDetail auditDetail, Object... args) {audit.log(message, AuditLevel.INFO, auditDetail, args);`

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/v3/OperationServiceBehavior.java [hunk 6]
```diff
@@ -303,12 +360,37 @@ public OperationUserActionResponse rejectOperation(OperationRejectRequest reques
             behavior.getCallbackUrlBehavior().notifyCallbackListenersOnOperationChange(savedEntity);
             final OperationDetailResponse operationDetailResponse = convertFromEntity(savedEntity);
 
+            logger.info("Operation rejected operation ID: {}, user ID: {}, application ID: {}.", operationId, userId, applicationId);
```
Comment (banterCZ) [inline powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/v3/OperationServiceBehavior.java:363]: I guest that the point `.` at the end is not needed.

**Review by banterCZ — APPROVED** [review]

File: docs/powerauth-server/PowerAuth-Server-1.3.0.md (hunk 3 above)
Comment (romanstrobl) [reply docs/powerauth-server/PowerAuth-Server-1.3.0.md:131]: We control the classes which call the auditing interface, it should be fine.

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/configuration/PowerAuthAuditConfiguration.java (hunk 4 above)
Comment (romanstrobl) [reply powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/configuration/PowerAuthAuditConfiguration.java:33]: Let's keep this change outside of current release.

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/v3/OperationServiceBehavior.java [hunk 7]
```diff
@@ -350,6 +432,17 @@ public OperationUserActionResponse failApprovalOperation(OperationFailApprovalRe
             behavior.getCallbackUrlBehavior().notifyCallbackListenersOnOperationChange(savedEntity);
             final OperationDetailResponse operationDetailResponse = convertFromEntity(savedEntity);
 
+            logger.info("Operation approval failed via explicit server call for operation ID: {}.", operationId);
```
Comment (romanstrobl) [inline powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/v3/OperationServiceBehavior.java:435]: dtto as above

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/v3/OperationServiceBehavior.java [hunk 8]
```diff
@@ -364,6 +457,17 @@ public OperationUserActionResponse failApprovalOperation(OperationFailApprovalRe
             behavior.getCallbackUrlBehavior().notifyCallbackListenersOnOperationChange(savedEntity);
             final OperationDetailResponse operationDetailResponse = convertFromEntity(savedEntity);
 
+            logger.info("Operation approval permanently failed via explicit server call for operation ID: {}.", operationId);
```
Comment (romanstrobl) [inline powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/v3/OperationServiceBehavior.java:460]: dtto as above

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/v3/OperationServiceBehavior.java [hunk 9]
```diff
@@ -398,6 +502,18 @@ public OperationDetailResponse cancelOperation(OperationCancelRequest request) t
 
         final OperationEntity savedEntity = operationRepository.save(operationEntity);
         behavior.getCallbackUrlBehavior().notifyCallbackListenersOnOperationChange(savedEntity);
+
+        logger.info("Operation canceled via explicit server call for operation ID: {}.", operationId);
```
Comment (romanstrobl) [inline powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/v3/OperationServiceBehavior.java:506]: dtto as above

File: docs/powerauth-server/PowerAuth-Server-1.3.0.md (hunk 2 above)
Comment (petrdvorak) [reply docs/powerauth-server/PowerAuth-Server-1.3.0.md:121]: Yes, the tables are from our separate auditing projects... We will need to explain this to the users, having "universal anything" is issue, because you either do not document it anywhere, or you copy paste to every repository.
